### DISCUSSION TITLE
Need help resolving "(HANDSHAKE_FAILURE): Received fatal alert: handshake_failure"

### DISCUSSION THREAD

**Discussion by ilitkevich** [ercot/api-specs#64] [Q&A]
Hi, 

I'm trying to establish a test connection to https://testmisapi.ercot.com/2007-08/Nodal/eEDS/EWS/ and got stuck on handshake_failure.  

I cloned https://github.com/ercot/ews-client and tried to make it work, which was not easy. I have followed the instruction from the repo and have truststore.jks with imported GlobalSignRootR3.cer and GlobalSignRSAOrganizationValidationCA2018.cer, and clientcert.jks with imported certificate from ERCOT .pfx file. 

Any ideas?

JDK 1.8.0_432

Arguments:
`-Djavax.net.ssl.trustStoreType=JKS
-Djavax.net.ssl.trustStore=paht\truststore.jks
-Djavax.net.ssl.trustStorePassword=changeit
-Dhttps.protocols=TLSv1.2,TLSv1.1,TLSv1,SSLv3
-jar path\ews-client-0.0.1-SNAPSHOT.jar`

System.setProperty("javax.net.debug", "ssl")  output:

`javax.net.ssl|FINE|01|main|2024-11-13 10:17:56.593 MST|Utilities.java:73|the previous server name in SNI (type=host_name (0), value=testmisapi.ercot.com) was replaced with (type=host_name (0), value=testmisapi.ercot.com)
javax.net.ssl|FINE|01|main|2024-11-13 10:17:56.593 MST|HandshakeContext.java:297|Ignore unsupported cipher suite: TLS_AES_256_GCM_SHA384 for TLS12
javax.net.ssl|FINE|01|main|2024-11-13 10:17:56.593 MST|HandshakeContext.java:297|Ignore unsupported cipher suite: TLS_AES_128_GCM_SHA256 for TLS12
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for EC
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for EC
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for EC
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for RSA
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for RSA
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for RSA
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for RSA
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for RSA
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for RSA
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for EC
javax.net.ssl|ALL|01|main|2024-11-13 10:17:56.719 MST|X509Authentication.java:215|No X.509 cert selected for RSA
javax.net.ssl|FINE|01|main|2024-11-13 10:17:56.800 MST|Alert.java:238|Received alert message (
"Alert": {
  "level"      : "fatal",
  "description": "handshake_failure"
}`

**Comment by CaffeineLab** [comment]
Did you import the ERCOT CA into the keystore as well?

**Reply by ilitkevich** [reply]
@CaffeineLab 

In my keystore clientcert.jks I have only my private certificate that I imported from ERCOT's .pfx like so:

`keytool –importkeystore –srckeystore yourcertificate.pfx –srcstoretype pkcs12 –srcstorepass yourcertificatepassword –srcalias yourcertificatealias –destkeystore clientcert.jks –deststoretype jks –deststorepass changeit –destalias clientcert`

In my truststore I have these two:
![image](https://github.com/user-attachments/assets/c7ff4fa6-1208-4903-87d5-565964dff9ea)

Do I need to add ERCOT CA to clientcert.jks?

**Reply by CaffeineLab** [reply]
It's been soooo long since I've messed with this, but looking at my scripts for importing things into the JKS, yes, I do import the ERCOT CA as well.  Should be a quick thing to test.

**Reply by ilitkevich** [reply]
@CaffeineLab unfortunately, it did not help. Same error message.

Keystore:

`Keystore type: jks
Keystore provider: SUN

Your keystore contains 2 entries

clientcert, 7-Nov-2024, PrivateKeyEntry,
Certificate fingerprint (SHA-256): DE:20:F4:32:12:B8:5D
ercotca, 13-Nov-2024, trustedCertEntry,
Certificate fingerprint (SHA-256): 2A:7D:9E:B0:0C:E3:4E`

Here's my truststore:

`Creation date: 4-Nov-2024
Entry type: trustedCertEntry

Owner: CN=GlobalSign RSA OV SSL CA 2018, O=GlobalSign nv-sa, C=BE
Issuer: CN=GlobalSign, O=GlobalSign, OU=GlobalSign Root CA - R3

...

Creation date: 4-Nov-2024
Entry type: trustedCertEntry

Owner: CN=GlobalSign, O=GlobalSign, OU=GlobalSign Root CA - R3
Issuer: CN=GlobalSign, O=GlobalSign, OU=GlobalSign Root CA - R3
`

**Reply by CaffeineLab** [reply]
We aren't apples to apples, but we have different protocol options.

Yours:
`-Djavax.net.ssl.trustStoreType=JKS -Djavax.net.ssl.trustStore=paht\truststore.jks -Djavax.net.ssl.trustStorePassword=changeit -Dhttps.protocols=TLSv1.2,TLSv1.1,TLSv1,SSLv3 -jar path\ews-client-0.0.1-SNAPSHOT.jar`

Ours (not complete, just the protocol options):
`-Dhttps.protocols=TLSv1.2 -Djdk.tls.client.protocols=TLSv1.2`

 
And other than that, I am a bit concerned about your misspelled 'paht' and the trustStorePasswordis still 'changeit', but if they are accurate, that's fine.
